### PR TITLE
Normalize nested fields

### PR DESCRIPTION
**Pull request by jfairbank**
This is my first hacky stab at fixing #516 after I realized I needed this functionality yesterday :)

The code has some repetitiveness and isn't the cleanest at the moment. All current tests pass, and I added some minimal nested fields to existing tests to ensure they work. Regardless, we might want some more tests. I also made sure to replicate the work done in #562 recently to preserve non-enumerable keys, but we might want tests for that as well.

I took the approach of nested normalizers instead of using the nested key sytax (e.g. `foo[].bar`). I found it easier to walk down the state and normalizer objects at the same time with this approach. Using the nested key syntax was a little more difficult.

For an idea of the syntax here is a test I changed:

```js
it('should normalize simple form values', () => {
  const defaultFields = {
    _active: undefined,
    _asyncValidating: false,
    [globalErrorKey]: undefined,
    _initialized: false,
    _submitting: false,
    _submitFailed: false
  };
  const normalize = reducer.normalize({
    foo: {
      name: () => 'normalized',
      person: {
        name: (name) => name && name.toUpperCase()
      },
      'pets[]': {
        name: (name) => name && name.toLowerCase()
      }
    }
  });
  const state = normalize({
    foo: {
      name: {
        value: 'dog'
      },
      person: {
        name: {
          value: 'John Doe',
        }
      },
      pets: [
        { name: { value: 'Fido' } },
        { name: { value: 'Tucker' } }
      ]
    }
  });
  expect(state)
    .toExist()
    .toBeA('object');
  expect(state.foo)
    .toExist()
    .toBeA('object')
    .toEqual({
      ...defaultFields,
      name: {
        value: 'normalized'
      },
      person: {
        name: {
          value: 'JOHN DOE'
        }
      },
      pets: [
        { name: { value: 'fido' } },
        { name: { value: 'tucker' } }
      ]
    });
});
```

One thing I'm still figuring out is how to handle this array situation:

```js
const normalize = reducer.normalize({
  // Array contains string field values
  'hobbies[]': (value) => value && value.toUpperCase()

  // Or array IS the field value
  'hobbies[]': (array) => array && array.map(hobby => hobby.toUpperCase())

  // Or array contains nested fields
  'people[]': (person) => {
    if (person) {
      return { name: person.name.toUpperCase() };
    }
    return { name: '' };
  }
});
```

Basically, if you don't specify nested properties, do you assume the array is a field value or that it still contains nested field values? What happens if the nested values have nested properties of their own? What would you expect to be the use case for this syntax? Or don't support it? If those questions don't make sense, let me know. Right now, I just have it assume that the array itself is the field value if you don't specify nested properties.

Please let me know if this is what you would like to see for normalizing nested fields and what I need to modify with this implementation.

### CODE REVIEW

**Comment by jfairbank** [comment]
Woops. I forgot I temporarily disabled linting when I was testing. Let me fix that :disappointed:

**Comment by erikras** [comment]
I think I vote for
```js
'hobbies[]': (array) => array && array.map(hobby => hobby.toUpperCase())
```
It's not hard to wrap your normalizing function in an `array.map()`. And this also allows you to do stuff like limiting array length or sorting or something.

How about this format? So that the keys match up exactly with the `fields` array syntax.
```js
'people[].name': name => name.toUpperCase()
```

**Comment by jfairbank** [comment]
> I think I vote for
> 
> `'hobbies[]': (array) => array && array.map(hobby => hobby.toUpperCase())`

That works for me. Seems least surprising.

> How about this format? So that the keys match up exactly with the fields array syntax.
> 
> `'people[].name': name => name.toUpperCase()`

Last night when I started fiddling with this, that syntax was not as easy to work with, but it may have been I was more tired at the end of my day ;)

The nested syntax which I did implement made it easier for walking down nested state even for arrays:

```js
normalize({
  foo: {
    'people[]': {
      name: (name) => name && name.toUpperCase()
    }
  }
});
```

This doesn't mean the fields syntax `'people[].name'` isn't doable, but I was also worried if I would have to walk the normalizer object more times. For example, if you wanted to specify `'people[].firstName'` and `'people[].lastName'`, then you would re-walk the state from the top. With the nested `'people[]': { firstName: ..., lastName: ... }`, you keep the context of where you are at as you walk down the state. If I'm not explaining that well, let me know.

That being said, I'd be happy to use what I've got to build out support for fields syntax (i.e. `'people[].name'`) because I like that mapping to the original fields syntax better too.

**Comment by erikras** [comment]
There a [many examples](https://github.com/erikras/redux-form/search?utf8=✓&q=openIndex) of tree crawling via the field string. Make sure you test all the possibilities, because, while it's not complicated, it's easy to make mistakes.

**Comment by jfairbank** [comment]
You're right, walking the tree wasn't so bad :sweat_smile:. My mind must've been exhausted or going down the wrong path the other night. I've switched the function over to support the field syntax (e.g. `'people[].name'`) instead. Here is the one test rewritten with the new syntax:

```js
it('should normalize simple form values', () => {
  const defaultFields = {
    _active: undefined,
    _asyncValidating: false,
    [globalErrorKey]: undefined,
    _initialized: false,
    _submitting: false,
    _submitFailed: false
  };
  const normalize = reducer.normalize({
    foo: {
      'name': () => 'normalized',
      'person.name': (name) => name && name.toUpperCase(),
      'pets[].name': (name) => name && name.toLowerCase()
    }
  });
  const state = normalize({
    foo: {
      name: {
        value: 'dog'
      },
      person: {
        name: {
          value: 'John Doe',
        }
      },
      pets: [
        { name: { value: 'Fido' } },
        { name: { value: 'Tucker' } }
      ]
    }
  });
  expect(state)
    .toExist()
    .toBeA('object');
  expect(state.foo)
    .toExist()
    .toBeA('object')
    .toEqual({
      ...defaultFields,
      name: {
        value: 'normalized'
      },
      person: {
        name: {
          value: 'JOHN DOE'
        }
      },
      pets: [
        { name: { value: 'fido' } },
        { name: { value: 'tucker' } }
      ]
    });
});
```

I also added a new test for arbitrarily deeply nested fields, one with and one without arrays. Are there other types of tests you would like to see?

(FYI, I'm piling commits on until we get to a satisfactory merge state and then I'll squash.)

**Comment by erikras** [comment]
Looks pretty good to me! :+1: :shipit:

**Comment by jfairbank** [comment]
Cool. I made one tiny tweak to remove an unnecessary variable, but it's all squashed and ready unless you want any more tests or tweaks.
